drivers/sensor/bmi08x/bmi08x_gyro.c

```diff
@@ -0,0 +1,472 @@
+}
+
+int bmi08x_gyro_reg_field_update(const struct device *dev, uint8_t reg_addr, uint8_t pos,
+				 uint8_t mask, uint8_t val)
+{
+	uint8_t old_val;
+
+	if (bmi08x_gyro_byte_read(dev, reg_addr, &old_val) < 0) {
+		return -EIO;
+	}
+
+	return bmi08x_gyro_byte_write(dev, reg_addr, (old_val & ~mask) | ((val << pos) & mask));
+}
+
+static const struct bmi08x_range bmi08x_gyr_range_map[] = {
+	{2000, BMI08X_GYR_RANGE_2000DPS}, {1000, BMI08X_GYR_RANGE_1000DPS},
+	{500, BMI08X_GYR_RANGE_500DPS},	  {250, BMI08X_GYR_RANGE_250DPS},
+	{125, BMI08X_GYR_RANGE_125DPS},
+};
+#define BMI08X_GYR_RANGE_MAP_SIZE ARRAY_SIZE(bmi08x_gyr_range_map)
+
+int32_t bmi08x_gyr_reg_val_to_range(uint8_t reg_val)
+{
+	return bmi08x_reg_val_to_range(reg_val, bmi08x_gyr_range_map, BMI08X_GYR_RANGE_MAP_SIZE);
+}
+
+static int bmi08x_gyr_odr_set(const struct device *dev, uint16_t freq_int, uint16_t freq_milli)
+{
+	int odr = bmi08x_freq_to_odr_val(freq_int, freq_milli);
+
+	if (odr < 0) {
+		return odr;
+	}
+
+	if (odr < BMI08X_GYRO_BW_532_ODR_2000_HZ || odr > BMI08X_GYRO_BW_32_ODR_100_HZ) {
+		return -ENOTSUP;
+	}
+
+	return bmi08x_gyro_byte_write(dev, BMI08X_REG_GYRO_BANDWIDTH, (uint8_t)odr);
+}
+
+static int bmi08x_gyr_range_set(const struct device *dev, uint16_t range)
+{
+	struct bmi08x_gyro_data *bmi08x = dev->data;
+	int32_t reg_val =
+		bmi08x_range_to_reg_val(range, bmi08x_gyr_range_map, BMI08X_GYR_RANGE_MAP_SIZE);
+
+	if (reg_val < 0) {
+		return reg_val;
+	}
+
+	if (bmi08x_gyro_byte_write(dev, BMI08X_REG_GYRO_RANGE, reg_val) < 0) {
+		return -EIO;
+	}
+
+	bmi08x->scale = BMI08X_GYR_SCALE(range);
+
+	return 0;
+}
+
+static int bmi08x_gyr_config(const struct device *dev, enum sensor_channel chan,
+			     enum sensor_attribute attr, const struct sensor_value *val)
+{
+	switch (attr) {
+	case SENSOR_ATTR_FULL_SCALE:
+		return bmi08x_gyr_range_set(dev, sensor_rad_to_degrees(val));
+	case SENSOR_ATTR_SAMPLING_FREQUENCY:
+		return bmi08x_gyr_odr_set(dev, val->val1, val->val2 / 1000);
+	default:
+		LOG_ERR("Gyro attribute not supported.");
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int bmi08x_attr_set(const struct device *dev, enum sensor_channel chan,
+			   enum sensor_attribute attr, const struct sensor_value *val)
+{
+#ifdef CONFIG_PM_DEVICE
+	enum pm_device_state state;
+
+	(void)pm_device_state_get(dev, &state);
+	if (state != PM_DEVICE_STATE_ACTIVE) {
+		return -EBUSY;
+	}
+#endif
+
+	switch (chan) {
+	case SENSOR_CHAN_GYRO_X:
+	case SENSOR_CHAN_GYRO_Y:
+	case SENSOR_CHAN_GYRO_Z:
+	case SENSOR_CHAN_GYRO_XYZ:
+		return bmi08x_gyr_config(dev, chan, attr, val);
+	default:
+		LOG_ERR("attr_set() not supported on this channel.");
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int bmi08x_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct bmi08x_gyro_data *bmi08x = dev->data;
+	size_t i;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+	if (bmi08x_gyro_transceive(dev, BMI08X_REG_GYRO_X_LSB | (1 << 7), false, bmi08x->gyr_sample,
```

Comment:
BMI08X_REG_GYRO_X_LSB | (1 << 7)

RATE_X_LSB address is 0x02, and I do not know why (0x02 | (1 << 7)) = 0x82?

Reply:
In transceive_spi you can add reg address with  (uint8_t addr = reg & 0x7f for write uint8_t addr = reg | 0x80 for read.)

Reply:
I can definitely tell English is not your first language here, but I'm not quite sure what you're trying to say here

bit 7 is to indicate is for a read, I probably could make a more clear by adding a macro
```
#define REG_READ_BIT BIT(7)
```

Reply:
> I can definitely tell English is not your first language here, but I'm not quite sure what you're trying to say here
> 
> bit 7 is to indicate is for a read, I probably could make a more clear by adding a macro
> 
> ```c
> #define REG_READ_BIT BIT(7)
> ```

what I means is operate register address or slave address must in the low level api not here.

Reply:
Your comment still doesn't make sense... but taking a look at this again, I should just replace this with the `bmi08x_gyro_read` function instead

Reply:
> Your comment still doesn't make sense to me and there is no requirement for it to be in the low level api... but taking a look at this again, I should just replace this with the `bmi08x_gyro_read` function instead

Let me put it this way:
In I2C mode, control read/write bit is in slave address bit7.
In SPI mode, control read/write bit is in register address bit0.

slave address is different with register address. `BMI08X_REG_GYRO_X_LSB` is a register address.

Reply:
You still haven't reviewed I2C yet... please do so.

Reply:
The 7b I2C address is always the same.... I know you're talking about the rw bit, but that is handled by the i2c API

Reply:
I see. you have changed it, this commit can be closed
```
	if (bmi08x_accel_read(dev, BMI08X_REG_ACCEL_X_LSB, (uint8_t *)data->acc_sample,
			      sizeof(data->acc_sample)) < 0) {
		pm_device_busy_clear(dev);
		return -EIO;
	}
```

and
```
	if (bmi08x_gyro_read(dev, BMI08X_REG_GYRO_X_LSB, (uint8_t *)bmi08x->gyr_sample,
			     sizeof(bmi08x->gyr_sample)) < 0) {
		return -EIO;
	}
```